sympy/assumptions/cnf.py

```diff
@@ -0,0 +1,373 @@
+from itertools import combinations, product
+
+from sympy import S
+from sympy.logic.boolalg import BooleanFunction, Or, And, Not
+
+
+# Early encoding for literals
+literals_store = dict()
+
+
+class Literal(object):
+    """
+    The smallest element of a CNF object
+    """
+    literals_count = 0
+
+    def __new__(cls, lit, is_not=False):
+        if type(lit).__name__ == 'Not':
+            lit = lit.args[0]
+            is_not = True
+        k = literals_store.get(lit, None)
+        if k is not None:
+            code = k
+        elif lit in (S.false, False):
+            code = 0
+        else:
+            cls.literals_count += 1
+            code = cls.literals_count
+        obj = super(Literal, cls).__new__(cls)
+        obj.code = code
+        obj.lit = lit
+        literals_store[lit] = obj.code
+        obj.is_Not = is_not
+        return obj
+
+    @property
+    def arg(self):
+        sign = -1 if self.is_Not else 1
+        return sign*self.code
+
+    def __invert__(self):
+        is_not = not self.is_Not
+        return Literal(self.lit, is_not=is_not)
+
+    def __str__(self):
+        return '%s( %s, %s)' % (type(self).__name__, self.lit, self.is_Not)
+
+    __repr__ = __str__
+
+    def __eq__(self, other):
+        return self.arg == other.arg
+
+    def __hash__(self):
+        h = hash((type(self).__name__, + self.arg))
+        return h
+
+    def __lt__(self, other):
```

Comment:
Where is this used?

Reply:
In the sorting of Literals. When two same literals are present, it is required for sorting them.

Reply:
It shouldn't be needed if you provide a sort key (as seems to be defined below). Does this `__lt__` actually make sense mathematically?

Reply:
In case two literals are equivalent, the sort key doesn't work. `__lt__` is needed for sorting then. Currently, it just compares the encoded form of literals ( e.g. `~Q.positive` is supposedly given an encoding of `-1` and that is compared and that is compared to others').

Reply:
Why doesn't it work if they are equivalent?